lib/reporters/base-reporter.js

```diff
@@ -0,0 +1,1005 @@
+'use strict';
+const os = require('os');
+const path = require('path');
+const stream = require('stream');
+
+const cliCursor = require('cli-cursor');
+const figures = require('figures');
+const indentString = require('indent-string');
+const ora = require('ora');
+const plur = require('plur');
+const prettyMs = require('pretty-ms');
+const trimOffNewlines = require('trim-off-newlines');
+const beautifyStack = require('./beautify-stack');
+
+const chalk = require('../chalk').get();
+const codeExcerpt = require('../code-excerpt');
+const colors = require('./colors');
+const formatSerializedError = require('./format-serialized-error');
+const improperUsageMessages = require('./improper-usage-messages');
+const prefixTitle = require('./prefix-title');
+const whileCorked = require('./while-corked');
+
+const nodeInternals = require('stack-utils').nodeInternals();
+
+class LineWriter extends stream.Writable {
+	constructor(dest, spinner) {
+		super();
+
+		this.dest = dest;
+		this.columns = dest.columns || 80;
+		this.lastLineIsEmpty = false;
+		this.lastSpinnerText = '';
+		this.spinner = spinner;
+	}
+
+	_write(chunk, encoding, callback) {
+		if (this.spinner) {
+			this.spinner.clear();
+			this._writeWithSpinner(chunk.toString('utf8'));
+		} else {
+			this.dest.write(chunk);
+		}
+
+		callback();
+	}
+
+	_writev(pieces, callback) {
+		if (!this.spinner) {
+			return;
+		}
+
+		// Discard the current spinner output. Any lines that were meant to be
+		// preserved should be rewritten.
+		this.spinner.clear();
+
+		const last = pieces.pop();
+		for (const piece of pieces) {
+			this.dest.write(piece.chunk);
+		}
+
+		this._writeWithSpinner(last.chunk.toString('utf8'));
+		callback();
+	}
+
+	_writeWithSpinner(string) {
+		if (!this.spinner) {
+			return;
+		}
+
+		if (!this.spinner.id) {
+			this.dest.write(string);
+			return;
+		}
+
+		this.lastSpinnerText = string;
+		// Ignore whitespace at the end of the chunk. We're continiously rewriting
+		// the last line through the spinner. Also be careful to remove the indent
+		// as the spinner adds its own.
+		this.spinner.text = string.trimEnd().slice(2);
+		this.spinner.render();
+	}
+
+	writeLine(string) {
+		if (string) {
+			this.write(indentString(string, 2) + os.EOL);
+			this.lastLineIsEmpty = false;
+		} else {
+			this.write(os.EOL);
+			this.lastLineIsEmpty = true;
+		}
+	}
+
+	ensureEmptyLine() {
+		if (!this.lastLineIsEmpty) {
+			this.writeLine();
+		}
+	}
+}
+
+class BaseReporter {
+	constructor(options) {
+		this.mode = options.mode;
+		this.reportStream = options.reportStream;
+		this.stdStream = options.stdStream;
+		this.watching = options.watching;
+		this.relativeFile = file => path.relative(options.projectDir, file);
+		this.consumeStateChange = whileCorked(
+			this.reportStream,
+			this.consumeStateChange
+		);
+
+		if (this.mode === 'verbose') {
```

Comment:
Maybe this should be a boolean, rather than a string?

Reply:
Yes, this is a good idea. I have been thinking about it too but I just wasn't quite sure if it would be too much change.